llvm/lib/CodeGen/MachineLateInstrsCleanup.cpp

```diff
@@ -225,8 +231,8 @@ bool MachineLateInstrsCleanup::processBlock(MachineBasicBlock *MBB) {
         MBBDefs.erase(Reg);
         MBBKills.erase(Reg);
       } else if (MI.findRegisterUseOperandIdx(Reg, TRI, true /*isKill*/) != -1)
-        // Keep track of register kills.
-        MBBKills[Reg] = &MI;
+        // Keep track of all instructions that fully or partially kills Reg.
+        MBBKills[Reg].push_back(&MI);
```

Comment:
You could also just do MRI.clearKillFlags any time it could matter

Reply:
That would be a lot simpler, just need to be sure it doesn't matter anywhere.